File: native/cpp/vector_ops.cpp
#include "vector_ops.h"
#include <cmath>
#include <cstdint>

namespace void_native {
static std::uint64_t hash_text(const std::string& text, unsigned int dimension) {
    std::uint64_t hash = 1469598103934665603ULL ^ dimension;
    for (unsigned char byte : text) { hash ^= byte; hash *= 1099511628211ULL; }
    return hash;
}

std::vector<double> assign_vector(const std::string& text, unsigned int dimensions) {
    std::vector<double> result(dimensions);
    double length = 0.0;
    for (unsigned int index = 0; index < dimensions; ++index) {
        result[index] = static_cast<double>(hash_text(text, index) % 2000001ULL) / 1000000.0 - 1.0;
        length += result[index] * result[index];
    }
    length = std::sqrt(length);
    if (length == 0.0) length = 1.0;
    for (double& value : result) value /= length;
    return result;
}

std::vector<double> normalize_weights(const std::vector<double>& values) {
    std::vector<double> result(values.size());
    double total = 0.0;
    for (double value : values) total += std::abs(value);
    if (total == 0.0) total = 1.0;
    for (std::size_t i = 0; i < values.size(); ++i) result[i] = std::abs(values[i]) / total;
    return result;
}

std::vector<double> blend_vectors(const std::vector<std::vector<double>>& vectors) {
    if (vectors.empty()) return {};
    std::vector<double> result(vectors.front().size(), 0.0);
    for (const auto& vector : vectors)
        for (std::size_t i = 0; i < result.size() && i < vector.size(); ++i) result[i] += vector[i];
    for (double& value : result) value /= static_cast<double>(vectors.size());
    return result;
}
}

File: native/rust/src/vector_ops.rs
pub fn assign_vector(text: &[u8], dimensions: usize) -> Vec<f64> {
    let mut values = Vec::with_capacity(dimensions);
    for dimension in 0..dimensions {
        let mut hash = 1469598103934665603_u64 ^ dimension as u64;
        for byte in text { hash ^= *byte as u64; hash = hash.wrapping_mul(1099511628211); }
        values.push((hash % 2_000_001) as f64 / 1_000_000.0 - 1.0);
    }
    let length = values.iter().map(|value| value * value).sum::<f64>().sqrt().max(1.0);
    values.into_iter().map(|value| value / length).collect()
}

pub fn blend(vectors: &[Vec<f64>]) -> Vec<f64> {
    if vectors.is_empty() { return Vec::new(); }
    let mut result = vec![0.0; vectors[0].len()];
    for vector in vectors { for (index, value) in vector.iter().enumerate() { result[index] += value; } }
    result.iter_mut().for_each(|value| *value /= vectors.len() as f64);
    result
}

pub fn norm(values: &[f64]) -> f64 { values.iter().map(|value| value * value).sum::<f64>().sqrt() }
pub fn dot(left: &[f64], right: &[f64]) -> f64 { left.iter().zip(right).map(|(a, b)| a * b).sum() }
pub fn same_dimensions(left: &[f64], right: &[f64]) -> bool { left.len() == right.len() }
pub fn cosine(left: &[f64], right: &[f64]) -> f64 { let denominator = norm(left) * norm(right); if denominator == 0.0 { 0.0 } else { dot(left, right) / denominator } }
pub fn scale(values: &[f64], factor: f64) -> Vec<f64> { values.iter().map(|value| value * factor).collect() }
pub fn add(left: &[f64], right: &[f64]) -> Vec<f64> { left.iter().zip(right).map(|(a, b)| a + b).collect() }
pub fn subtract(left: &[f64], right: &[f64]) -> Vec<f64> { left.iter().zip(right).map(|(a, b)| a - b).collect() }
pub fn distance(left: &[f64], right: &[f64]) -> f64 { norm(&subtract(left, right)) }
pub fn dominant(values: &[f64]) -> usize { values.iter().enumerate().max_by(|(_, a), (_, b)| a.abs().total_cmp(&b.abs())).map(|(index, _)| index).unwrap_or(0) }
pub fn finite(values: &[f64]) -> bool { values.iter().all(|value| value.is_finite()) }
pub fn summary(values: &[f64]) -> String { format!("dimensions={}, norm={}, dominant={}, finite={}", values.len(), norm(values), dominant(values), finite(values)) }
